README.md
# react-native-gplaces

React Native library for utilizing Google's Autocomplete for Places.

## Install

**Notice**: Geolocation must be linked, in React Native 0.60+ this is done automatically.

### using npm

```sh
npm install --save react-native-gplaces @react-native-community/geolocation
```

### using yarn

```sh
yarn add react-native-gplaces @react-native-community/geolocation
```

## Importing

### ES6

The module uses an ES6 style export statement, simply use `import` to load the module.

```js
import GPlaces from 'react-native-gplaces';
```

### ES5

If you're using an ES5 require statement to load the module, please add `default`. Look [here](https://github.com/joltup/rn-fetch-blob/wiki/Trouble-Shooting#rnfetchblobfetch-is-not-a-function) for more detail.

```js
var GPlaces = require('react-native-gplaces').default;
```

## Usage

This package supports custom queries if you'd like to use them. For searching queries look [here](https://developers.google.com/places/web-service/autocomplete#place_autocomplete_requests) and queries for getting places look [here](https://developers.google.com/places/web-service/details#PlaceDetailsRequests).

### Creating an instance

You can obtain a Google API key [here](https://developers.google.com/maps/documentation/javascript/get-api-key).

```js
const places = new GPlaces({
  key: 'YOUR_GOOGLE_MAPS_API_KEY' // https://developers.google.com/maps/documentation/javascript/get-api-key
});
```

### Search for places

```js
places.search('nemo', {
  components: 'country:nl',
  types: 'establishment'
})
  .then(r => {
    // returns ACResult[]
  })
  .catch(console.error)
```

### Search for places nearby

```js
places.searchNearby('brussel', 2500, {
  components: 'country:be',
  types: '(cities)'
})
  .then(r => {
    // returns ACResult[]
  })
  .catch(console.error)
```

### Get details for a Place

```js
places.getPlaceDetails('ChIJn8N5VRvZxkcRmLlkgWTSmvM', {
  fields: 'geometry'
})
  .then(r => {
    // returns PDResult
  })
  .catch(console.error)
```

## API

### `constructor GPlaces(options: Options): GPlaces`

Constructor for creating an instance of the GPlaces class

### `request: (url?: string) => Promise<any> (@private)`

Calling this method will fetch an URL and return a JSON response

### `autocompleteRequest: (options?: string) => Promise<ACResult[]> (@private)`

Calling this method will fetch and return an array of results

<!-- TODO: document default query -->
<!-- TODO: recommend debounce -->

### `search: (input?: string, query?: ACQuery | undefined) => Promise<ACResult[]>`

Calling this method will search for places matching the input.

### `searchNearby: (input?: string, radius?: number, query?: ACQuery | undefined) => Promise<ACResult[]>`

Calling this method will search for nearby places matching the input in a given radius.
The default radius is 1000m / 1km.

### `getPlaceDetails: (placeid?: string, query?: PDQuery | undefined) => Promise<PDResult>`

Calling this method will get certain details for a place based on a query.

## Types

For up-to-date information about Autocomplete queries look [here](https://developers.google.com/places/web-service/autocomplete#place_autocomplete_requests), for Place Details queries look [here](https://developers.google.com/places/web-service/details#PlaceDetailsRequests).

```ts
export interface ACQuery {
  sessiontoken?: string;
  offset?: number;
  location?: string;
  radius?: number;
  language?: string;
  types?: string;
  components?: string;
  strictbounds?: boolean;
}

export interface PDQuery {
  language?: string;
  region?: string;
  sessiontoken?: string;
  fields?: string;
}

export interface Options {
  key: string;
}

export interface MSubstring {
  length: number;
  offset: number;
}
export type MSubstrings = Array<MSubstring>

export interface SFormat {
  main_text: string;
  main_text_matched_substrings: MSubstrings;
  secondary_text: string;
}
export type SFormatting = Array<SFormat>

export interface Term {
  offset: number;
  value: string;
}
export type Terms = Array<Term>

export interface ACResult {
  id: string;
  place_id: string;
  reference: string;
  description: string;
  matched_substrings: MSubstrings;
  structured_formatting: SFormatting;
  terms: Terms;
  types: Array<string>;
}
export type ACResults = Array<ACResult>

export type Types = Array<string>
export type HTMLAttrs = Array<string>

export interface AddressComponent {
  long_name: string;
  short_name: string;
  types: Types;
}
export type AddressComponents = Array<AddressComponent>

export interface Photo {
  height: number;
  html_attributions: HTMLAttrs;
  photo_reference: string;
  width: number;
}
export type Photos = Array<Photo>

export interface Location {
  lat: string;
  lng: string;
}
export interface Viewport {
  northeast: Location;
  southwest: Location;
}
export interface Geometry {
  location: Location;
  viewport: Viewport;
}

export interface PDResult {
  address_components: AddressComponents;
  adr_address: string;
  formatted_address: string;
  geometry: Geometry;
  icon: string;
  id: string;
  name: string;
  photos: Photos;
  place_id: string;
  reference: string;
  scope: string;
  types: Types;
  url: string;
  utc_offset: number;
  vicinity: string;
}
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

**Notice**: Keeping a changelog since 3.0.0

## 3.0.0

Added:
- Support for React Native 0.60+
- `@react-native-community/geolocation` as a dependency

Changed:
- Updated peerDependencies

Fixed:
- Warning about Geolocation being moved from `react-native`


index.js
import { Platform } from 'react-native'
import Qs from 'qs'

interface Query {
  sessiontoken: String; // A random string which identifies an autocomplete session for billing purposes. If this parameter is omitted from an autocomplete request, the request is billed independently. See the pricing sheet for details.
  offset: Number; // The position, in the input term, of the last character that the service uses to match predictions. For example, if the input is 'Google' and the offset is 3, the service will match on 'Goo'. The string determined by the offset is matched against the first word in the input term only. For example, if the input term is 'Google abc' and the offset is 3, the service will attempt to match against 'Goo abc'. If no offset is supplied, the service will use the whole term. The offset should generally be set to the position of the text caret.
  location: String; // The point around which you wish to retrieve place information. Must be specified as "latitude,longitude".
  radius: Number; // The distance (in meters) within which to return place results. Note that setting a radius biases results to the indicated area, but may not fully restrict results to the specified area. See Location Biasing and Location Restrict below.
  language: String; // The language code, indicating in which language the results should be returned, if possible. Searches are also biased to the selected language; results in the selected language may be given a higher ranking. See the list of supported languages and their codes. Note that we often update supported languages so this list may not be exhaustive. If language is not supplied, the Place Autocomplete service will attempt to use the native language of the domain from which the request is sent.
  types: String; // The types of place results to return. See Place Types below. If no type is specified, all types will be returned.
  components: String; // A grouping of places to which you would like to restrict your results. Currently, you can use components to filter by up to 5 countries. Countries must be passed as a two character, ISO 3166-1 Alpha-2 compatible country code. For example: components=country:fr would restrict your results to places within France. Multiple countries must be passed as multiple country:XX filters, with the pipe character (|) as a separator. For example: components=country:us|country:pr|country:vi|country:gu|country:mp would restrict your results to places within the United States and its unincorporated organized territories.
  strictbounds: Boolean; // Returns only those places that are strictly within the region defined by location and radius. This is a restriction, rather than a bias, meaning that results outside this region will not be returned even if they match the user input.
}
interface Options {
  key: String; // Your application's API key. This key identifies your application for purposes of quota management. See Get a key for more information. Google Maps APIs Premium Plan customers must use the API project created for them as part of their Premium Plan purchase.
  query: Query; // Certain parameters are required to initiate a Place Autocomplete request. As is standard in URLs, all parameters are separated using the ampersand (&) character. The list of parameters and their possible values are enumerated above.
}

interface MSubstring {
  length: Number; // length of substring in text
  offset: Number; // position of substring in text
}
type MSubstrings = Array<MSubstring>

interface SFormat {
  main_text: String; // contains the main text of a prediction, usually the name of the place.
  main_text_matched_substrings: MSubstrings; // contains an array with offset value and length. These describe the location of the entered term in the prediction result text, so that the term can be highlighted if desired.
  secondary_text: String; // contains the secondary text of a prediction, usually the location of the place.
}
type SFormatting = Array<SFormat>

interface Term {
  offset: Number; // position of term in text
  value: String; // value of term in text
}
type Terms = Array<Term>

interface Result {
  id: String; // unique id for the search result
  place_id: String; // is a textual identifier that uniquely identifies a place. To retrieve information about the place, pass this identifier in the placeId field of a Places API request. For more information about place IDs, see the place ID overview.
  reference: String; // same as place_id
  description: String; // contains the human-readable name for the returned result. For establishment results, this is usually the business name
  matched_substrings: MSubstring; // contains an array with offset value and length. These describe the location of the entered term in the prediction result text, so that the term can be highlighted if desired.
  structured_formatting: SFormatting; // contains the following subfields: main_text, main_text_matched_substrings, secondary_text
  terms: Terms; // contains an array of terms identifying each section of the returned description (a section of the description is generally terminated with a comma). Each entry in the array has a value field, containing the text of the term, and an offset field, defining the start position of this term in the description, measured in Unicode characters.
  types: Array<String>; // contains an array of types that apply to this place. For example: [ "political", "locality" ] or [ "establishment", "geocode" ].
}
type Results = Array<Result>

/** Class containing methods to get results from Google Places */
// TODO: add nearby search without input
export default class GPlaces {
  options: Options
  /**
   * Create an instance of the GPlaces class.
   * @param {Options} options
   *        Options for creating the handler
   */
  constructor (options: Options) {
    this.options = options
  }

  /**
   * Make a request expecting a JSON response
   * @param {String} url
   *        Request URL for Google Maps
   * @private
   */
  request = (url: String) => new Promise((resolve, reject) => {
    if (!url || typeof url !== 'string') {
      reject(new Error('Invalid request URL.'))
    }

    fetch(url)
      .then(r => r.json())
      .then(resolve)
      .catch(reject)
  })

  /**
   * Search for places matching input
   * @param {String} input
   *        Input string for the search
   * @returns {Promise<Results>}
   *          Returns an unaltered array of results from the Google Places API
   */
  search = (input: String) => new Promise(async (resolve, reject) => {
    if (!input || typeof input !== 'string') {
      reject(new Error('Invalid input string given.'))
    }

    const options = Qs.stringify({
      input,
      language: 'en',
      key: this.options.key,
      types: 'establishment',
      ...this.options.query
    })

    const response = await this.request(
      `https://maps.googleapis.com/maps/api/place/autocomplete/json?${options}`
      // `https://maps.googleapis.com/maps/api/geocode/json?${options}`
    ).catch(reject)

    resolve((response || {}).predictions || [])
  })

  /**
   * Search for places matching input
   * @param {String} input
   *        Input string for the search
   * @param {Number} radius
   *        Radius in meters to look for places
   * @returns {Promise<Results>}
   *          Returns an unaltered array of results from the Google Places API
   */
  searchNearby = (input: String, radius: Number = 1000) => new Promise(async (resolve, reject) => {
    navigator.geolocation.getCurrentPosition(async position => {
      const { latitude, longitude } = (position || {}).coords || {}

      if (!latitude || !longitude) {
        reject(new Error('No latitude/longtitude for current location found.'))
      }

      if (!input || typeof input !== 'string') {
        reject(new Error('Invalid input string given.'))
      }

      if (!radius || typeof radius !== 'number') {
        reject(new Error('Invalid radius given.'))
      }

      const options = Qs.stringify({
        input,
        radius,
        language: 'en',
        key: this.options.key,
        types: 'establishment',
        strictbounds: true,
        location: `${latitude},${longitude}`,
        ...this.options.query
      })

      const response = await this.request(
        `https://maps.googleapis.com/maps/api/place/autocomplete/json?${options}`
        // `https://maps.googleapis.com/maps/api/geocode/json?${options}`
      ).catch(reject)

      resolve((response || {}).predictions || [])
    }, reject, {
      enableHighAccuracy: Platform.OS === 'android',
      timeout: 10000,
      maximumAge: 1000
    })
  })
}


src/types.ts
// TODO: add enums (for example: fields could have 'geometry')

/** Autocomplete Query */
export interface ACQuery {
  /** A random string which identifies an autocomplete session for billing purposes. If this parameter is omitted from an autocomplete request, the request is billed independently. See the pricing sheet for details. */
  sessiontoken?: string;
  /** The position, in the input term, of the last character that the service uses to match predictions. For example, if the input is 'Google' and the offset is 3, the service will match on 'Goo'. The string determined by the offset is matched against the first word in the input term only. For example, if the input term is 'Google abc' and the offset is 3, the service will attempt to match against 'Goo abc'. If no offset is supplied, the service will use the whole term. The offset should generally be set to the position of the text caret. */
  offset?: number;
  /** The point around which you wish to retrieve place information. Must be specified as "latitude,longitude". */
  location?: string;
  /** The distance (in meters) within which to return place results. Note that setting a radius biases results to the indicated area, but may not fully restrict results to the specified area. See Location Biasing and Location Restrict below. */
  radius?: number;
  /** The language code, indicating in which language the results should be returned, if possible. Searches are also biased to the selected language; results in the selected language may be given a higher ranking. See the list of supported languages and their codes. Note that we often update supported languages so this list may not be exhaustive. If language is not supplied, the Place Autocomplete service will attempt to use the native language of the domain from which the request is sent. */
  language?: string;
  /** The types of place results to return. See Place Types below. If no type is specified, all types will be returned. */
  types?: string;
  /** A grouping of places to which you would like to restrict your results. Currently, you can use components to filter by up to 5 countries. Countries must be passed as a two character, ISO 3166-1 Alpha-2 compatible country code. For example: components=country:fr would restrict your results to places within France. Multiple countries must be passed as multiple country:XX filters, with the pipe character (|) as a separator. For example: components=country:us|country:pr|country:vi|country:gu|country:mp would restrict your results to places within the United States and its unincorporated organized territories. */
  components?: string;
  /** Returns only those places that are strictly within the region defined by location and radius. This is a restriction, rather than a bias, meaning that results outside this region will not be returned even if they match the user input. */
  strictbounds?: boolean;
}

/** Place Details Query */
export interface PDQuery {
  /** The language code, indicating in which language the results should be returned, if possible. Note that some fields may not be available in the requested language. */
  language?: string;
  /** The region code, specified as a ccTLD (country code top-level domain) two-character value. Most ccTLD codes are identical to ISO 3166-1 codes, with some exceptions. This parameter will only influence, not fully restrict, results. If more relevant results exist outside of the specified region, they may be included. When this parameter is used, the country name is omitted from the resulting formatted_address for results in the specified region. */
  region?: string;
  /** A random string which identifies an autocomplete session for billing purposes. */
  sessiontoken?: string;
  /** One or more fields, specifying the types of place data to return, separated by a comma. */
  fields?: string;
}

/** API Options */
export interface Options {
  /** Your application's API key. This key identifies your application for purposes of quota management. See Get a key for more information. Google Maps APIs Premium Plan customers must use the API project created for them as part of their Premium Plan purchase. */
  key: string;
}

/** Matched Substring */
export interface MSubstring {
  /** length of substring in text */
  length: number;
  /** position of substring in text */
  offset: number;
}
/** Matched Substrings */
export type MSubstrings = Array<MSubstring>

/** Structured Format */
export interface SFormat {
  /** contains the main text of a prediction, usually the name of the place */
  main_text: string;
  /** contains an array with offset value and length */
  main_text_matched_substrings: MSubstrings;
  /** contains the secondary text of a prediction, usually the location of the place */
  secondary_text: string;
}
/** Structured Formatting */
export type SFormatting = Array<SFormat>

/** Term */
export interface Term {
  /** position of term in text */
  offset: number;
  /** value of term in text */
  value: string;
}
/** Terms */
export type Terms = Array<Term>

/** Autocomplete Result */
export interface ACResult {
  /** unique id for the search result */
  id: string;
  /** textual identifier that uniquely identifies a place */
  place_id: string;
  /** same as place_id */
  reference: string;
  /** contains the human-readable name for the returned result (for establishment results, this is usually the business name) */
  description: string;
  /** contains an array with offset value and length */
  matched_substrings: MSubstrings;
  /** contains the following subfields: main_text, main_text_matched_substrings, secondary_text */
  structured_formatting: SFormatting;
  /** contains an array of terms identifying each section of the returned description */
  terms: Terms;
  /** contains an array of types that apply to this place (for example: ["political", "locality"] or ["establishment", "geocode"]) */
  types: Array<string>;
}
/** Autocomplete Results */
export type ACResults = Array<ACResult>

/** Types */
export type Types = Array<string>
/** HTML Attributions */
export type HTMLAttrs = Array<string>

/** Address Component */
// TODO: add jsdoc
export interface AddressComponent {
  long_name: string;
  short_name: string;
  types: Types;
}
/** Address Components */
export type AddressComponents = Array<AddressComponent>

/** Photo */
// TODO: add jsdoc
export interface Photo {
  height: number;
  html_attributions: HTMLAttrs;
  photo_reference: string;
  width: number;
}
/** Address Components */
export type Photos = Array<Photo>

/** Location */
export interface Location {
  /** latitude of the location */
  lat: string;
  /** longitude of the location */
  lng: string;
}
/** Viewport */
export interface Viewport {
  /** north east location of the viewport */
  northeast: Location;
  /** south west location of the viewport */
  southwest: Location;
}
/** Geometry Details */
export interface Geometry {
  /** contains the geocoded latitude,longitude value for this place */
  location: Location;
  /** contains the preferred viewport when displaying this place on a map as a Location if it is known */
  viewport: Viewport;
}

/** Place Details Result */
export interface PDResult {
  /** an array containing the separate components applicable to this address */
  address_components: AddressComponents;
  /** a representation of the place's address in the adr microformat */
  adr_address: string;
  /** a string containing the human-readable address of this place */
  formatted_address: string;
  /** geodata of the place */
  geometry: Geometry;
  /** contains the URL of a suggested icon which may be displayed to the user when indicating this result on a map */
  icon: string;
  /** unique id for the search result */
  id: string;
  /** contains the human-readable name for the returned result (for establishment results, this is usually the canonicalized business name) */
  name: string;
  /** an array of Photo objects, each containing a reference to an image (a Place Details request may return up to ten photos) */
  photos: Photos;
  /** textual identifier that uniquely identifies a place */
  place_id: string;
  /** same as place_id */
  reference: string;
  /** indicates the scope of the place_id */
  scope: string;
  /** contains an array of feature types describing the given result */
  types: Types;
  /** contains the URL of the official Google page for this place */
  url: string;
  /** contains the number of minutes this place’s current timezone is offset from UTC */
  utc_offset: number;
  /** lists a simplified address for the place, including the street name, street number, and locality, but not the province/state, postal code, or country */
  vicinity: string;
}


src/index.ts
import { Platform } from 'react-native'
import Geolocation from '@react-native-community/geolocation'
import { GeolocationResponse } from '@react-native-community/geolocation/typings'
import * as Qs from 'qs'
import {
  ACQuery,
  PDQuery,
  Options,
  PDResult,
  ACResults
} from './types'

/** Class containing methods to get results from Google Places */
// TODO: add nearby search without input
export default class GPlaces {
  /** API options */
  options: Options
  /**
   * Create an instance of the GPlaces class.
   * @param   {Options} [options=]
   *          Options for creating the handler
   */
  // TODO: give error if no key
  constructor (options: Options) {
    if (!options) {
      this.options = { key: '' }
    } else {
      this.options = options
    }
  }

  /**
   * Make a request expecting a JSON response
   * @param   {string} [url='']
   *          Request URL for Google Maps
   * @private
   */
  request = (
    url: string = ''
  ) => new Promise<any>((resolve, reject) => {
    if (!url || typeof url !== 'string') {
      reject(new Error('Invalid request URL.'))
    }

    fetch(url)
      .then(r => r.json())
      .then(resolve)
      .catch(reject)
  })

  /**
   * Search for places using specified options
   * @param   {string} [options='']
   *          Options string for the API call
   * @returns {Promise<ACResults>}
   *          Returns an unaltered array of results from the Google Places API
   */
  autocompleteRequest = (
    options: string = ''
  ) => new Promise<ACResults>((resolve, reject) => {
    this.request(`https://maps.googleapis.com/maps/api/place/autocomplete/json?${options}`)
      .then(res => {
        if (!res || !res.predictions || !Array.isArray(res.predictions)) {
          return reject(new Error('Invalid response from server'))
        }
        resolve(res.predictions)
      })
      .catch(reject)
  })

  /**
   * Search for places matching input
   * @param   {string} [input='']
   *          Input string for the search
   * @param   {ACQuery} [query=]
   *          Query object for the search
   * @returns {Promise<ACResults>}
   *          Returns an unaltered array of results from the Google Places API
   */
  search = (
    input: string = '',
    query?: ACQuery
  ) => new Promise<ACResults>(async (resolve, reject) => {
    if (!input || typeof input !== 'string') {
      reject(new Error('Invalid input string given.'))
    }

    const options = Qs.stringify({
      input,
      language: 'en',
      key: this.options.key,
      types: 'establishment',
      ...query
    })

    this.autocompleteRequest(options)
      .then(resolve)
      .catch(reject)
  })

  /**
   * Search for places matching input
   * @param   {string} [input='']
   *          Input string for the search
   * @param   {number} [radius=1000]
   *          Radius in meters to look for places
   * @param   {ACQuery} [query=]
   *          Query object for the search
   * @returns {Promise<ACResults>}
   *          Returns an unaltered array of results from the Google Places API
   */
  searchNearby = (
    input: string = '',
    radius: number = 1000,
    query?: ACQuery
  ) => new Promise<ACResults>(async (resolve, reject) => {
    Geolocation.getCurrentPosition(async (position: GeolocationResponse) => {
      const { latitude = 0, longitude = 0 } = (position || { coords: null, timestamp: null }).coords || {}

      if (!latitude || !longitude) {
        reject(new Error('No latitude/longtitude for current location found.'))
      }

      if (!input || typeof input !== 'string') {
        reject(new Error('Invalid input string given.'))
      }

      if (!radius || typeof radius !== 'number') {
        reject(new Error('Invalid radius given.'))
      }

      const options = Qs.stringify({
        input,
        radius,
        language: 'en',
        key: this.options.key,
        types: 'establishment',
        strictbounds: true,
        location: `${latitude},${longitude}`,
        ...query
      })

      this.autocompleteRequest(options)
        .then(resolve)
        .catch(reject)
    }, reject, {
      enableHighAccuracy: Platform.OS === 'android',
      timeout: 10000,
      maximumAge: 1000
    })
  })

  /**
   * Search for places matching input
   * @param   {string} [placeid='']
   *          Place ID string of the requested place
   * @param   {PDQuery} [query=]
   *          Place ID string of the requested place
   * @returns {Promise<PDResult>}
   *          Returns an unaltered result object from the Google Places API
   */
  getPlaceDetails = (placeid: string = '', query?: PDQuery) => new Promise<PDResult>(async (resolve, reject) => {
    if (!placeid || typeof placeid !== 'string') {
      reject(new Error('Invalid Place ID string given.'))
    }

    const options = Qs.stringify({
      placeid,
      language: 'en',
      fields: 'geometry',
      key: this.options.key,
      ...query
    })

    this.request(`https://maps.googleapis.com/maps/api/place/details/json?${options}`)
      .then(res => {
        if (!res || !res.result || (typeof res !== 'object')) {
          return reject(new Error('Invalid response from server'))
        }
        resolve(res.result)
      })
      .catch(reject)
  })
}
